include/aos/common/crypto.hpp

```diff
@@ -0,0 +1,266 @@
+private:
+    StaticArray<uint8_t, cCertificatePubKeyLen>  mPublicKey;
+    StaticArray<uint8_t, cCertificatePrivKeyLen> mPrivateKey;
+};
+
+namespace asn1 {
+/**
+ * ASN.1 OBJECT IDENTIFIER
+ */
+using ObjectIdentifier = StaticString<cASN1ObjIdLen>;
+
+/**
+ * ASN.1 structure extension. RFC 580, section 4.2
+ */
+struct Extension {
+    ObjectIdentifier                        mId;
+    StaticArray<uint8_t, cCertExtValueSize> mValue;
+};
+
+/**
+ * ASN1 Encodes string into ASN1 value
+ */
+
+inline void Encode(const String& /*src*/, Array<uint8_t>& /*asn1Value*/)
+{
+}
+
+/**
+ *
+ */
+
+} // namespace asn1
+
+namespace x509 {
+
+/**
+ * x509 Certificate.
+ */
+struct Certificate {
+    /**
+     * Certificate subject.
+     */
+    StaticString<cCertificateSubjSize> mSubject;
+
+    /**
+     * Certificate subject key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mSubjectKeyId;
+
+    /**
+     * Certificate authority key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mAuthorityKeyId;
+
+    /**
+     * Certificate issuer name.
+     */
+    StaticString<cCertificateIssuerLen> mIssuer;
+    /**
+     * Certificate serial number.
+     */
+    StaticArray<uint8_t, cCertificateSerialNumberLen> mSerial;
+    /**
+     * Certificate validity period.
+     */
+    time::Time mNotBefore, mNotAfter;
+};
+
+/**
+ * Provides interface to manage certificates.
+ */
+class CertificateProviderItf {
+public:
+    /**
+     * Creates a new certificate based on a template.
+     *
+     * @param templ a pattern for a new certificate.
+     * @param parent a parent certificate in a certificate chain.
+     * @param pubKey public key.
+     * @param privKey private key.
+     * @param[out] resultCert result certificate in PEM format.
+     * @result Error.
+     */
+    virtual Error CreateCertificate(
+        const Certificate& templ, const Certificate& parent, const PrivateKey& privKey, Array<uint8_t>& pemCert)
+        = 0;
+
+    /**
+     * Creates certificates from a PEM blob.
+     *
+     * @param pemBlob raw certificate in a PEM format.
+     * @param[out] resultCerts result certificates.
+     * @result Error.
+     */
+    virtual Error CreateCertificatesFromPEM(const Array<uint8_t>& pemBlob, Array<Certificate>& resultCerts) = 0;
+
+    /**
+     * Destroys object instace.
+     */
+    virtual ~CertificateProviderItf() = default;
+};
+
+/**
+ * x509 Certificate request.
+ */
+struct CSR {
+    /**
+     * Raw CSR content.
+     */
+    StaticArray<uint8_t, cRawCSRSize> mRaw;
+    /**
+     * Certificate subject size.
+     */
+    StaticString<cCertificateSubjSize> mSubject;
+    /**
+     * Alternative DNS names.
+     */
+    StaticArray<StaticString<cDnsNameLen>, cAltDnsNamesCount> mDnsNames;
+    /**
+     * Contains extra extensions applied to CSR.
+     */
+    StaticArray<asn1::Extension, cCertificateExtraExtCount> mExtraExtensions;
+};
+
+/**
+ * Provides interface to manage certificate requests.
+ */
+class CSRProviderItf {
```

Comment:
No need different interface for each method. One `x509ProviderItf` would be enough.